File: frontend/src/types/listing.ts
import { Amenity } from "./meta";

export interface UserSummary {
  id: number;
  name: string;
  avatar_url?: string;
  is_superhost: boolean;
}

export interface ListingImage {
  id: number;
  url: string;
  position: number;
}

export interface ListingSummary {
  id: number;
  title: string;
  property_type: string;
  category: string;
  city: string;
  country: string;
  price_per_night: number;
  cleaning_fee: number;
  max_guests: number;
  bedrooms: number;
  beds: number;
  bathrooms: number;
  images: ListingImage[];
  amenities: Amenity[];
  host: UserSummary;
  rating?: number | null;
  reviews_count: number;
}

export interface Review {
  id: number;
  rating: number;
  comment: string;
  created_at: string;
  author: UserSummary;
}

export interface ListingDetail extends ListingSummary {
  description: string;
  latitude?: number | null;
  longitude?: number | null;
  created_at: string;
  updated_at: string;
  reviews?: Review[];
}

export interface PaginatedListings {
  items: ListingSummary[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface ListingSearchParams {
  location?: string;
  check_in?: string;
  check_out?: string;
  guests?: number;
  min_price?: number;
  max_price?: number;
  property_type?: string;
  category?: string;
  amenities?: string[];
  bedrooms?: number;
  beds?: number;
  page?: number;
  page_size?: number;
}

export interface UnavailableDateRange {
  check_in: string;
  check_out: string;
}

export interface QuoteResponse {
  listing_id: number;
  check_in: string;
  check_out: string;
  guests: number;
  nights: number;
  nightly_price: number;
  subtotal: number;
  cleaning_fee: number;
  service_fee: number;
  total: number;
  is_available: boolean;
}

export interface BookingResponse {
  id: number;
  listing_id: number;
  guest_id: number;
  check_in: string;
  check_out: string;
  guests: number;
  nightly_price: number;
  nights: number;
  cleaning_fee: number;
  service_fee: number;
  total_price: number;
  status: string;
  created_at: string;
  listing?: ListingSummary;
  guest?: UserSummary;
}

File: backend/app/schemas/listing.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, Field
from app.schemas.meta import AmenitySchema

class UserSummarySchema(BaseModel):
    id: int
    name: str
    avatar_url: Optional[str] = None
    is_superhost: bool

    class Config:
        from_attributes = True

class ListingImageSchema(BaseModel):
    id: int
    url: str
    position: int

    class Config:
        from_attributes = True

class ListingSummaryResponse(BaseModel):
    id: int
    title: str
    property_type: str
    category: str
    city: str
    country: str
    price_per_night: float
    cleaning_fee: float
    max_guests: int
    bedrooms: int
    beds: int
    bathrooms: float
    images: List[ListingImageSchema] = []
    amenities: List[AmenitySchema] = []
    host: UserSummarySchema
    rating: Optional[float] = None
    reviews_count: int = 0

    class Config:
        from_attributes = True

class ReviewSchema(BaseModel):
    id: int
    rating: int
    comment: str
    created_at: datetime
    author: UserSummarySchema

    class Config:
        from_attributes = True

class ListingDetailResponse(ListingSummaryResponse):
    description: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    created_at: datetime
    updated_at: datetime
    reviews: List[ReviewSchema] = []

    class Config:
        from_attributes = True

class PaginatedListingResponse(BaseModel):
    items: List[ListingSummaryResponse]
    page: int
    page_size: int
    total: int
    total_pages: int

class UnavailableDateRange(BaseModel):
    check_in: date
    check_out: date

    class Config:
        from_attributes = True

class QuoteResponse(BaseModel):
    listing_id: int
    check_in: date
    check_out: date
    guests: int
    nights: int
    nightly_price: float
    subtotal: float
    cleaning_fee: float
    service_fee: float
    total: float
    is_available: bool = True
